### PR TITLE
TMS-2128 support for TCP tunnels

### PR DESCRIPTION
**Pull request by rjeczalik**
This CL adds support for TCP tunnels, with support for either port-based or IP-based routing.

### CODE REVIEW

**Comment by cihangir** [comment]
hi @rjeczalik  can you please send refactoring and implementation in two PR, since this is an important feature, lets not have interesting stuff to sneak in :)

**Comment by rjeczalik** [comment]
@cihangir Done. In another PR I'm going to send tests + refactoring. There may be still old flakes in tests (like https://travis-ci.org/koding/tunnel/jobs/113488281), I'll take them down after both PR are merged.

File: client.go [hunk 1]
```diff
@@ -384,21 +432,21 @@ func (c *Client) proxy(port int) error {
...
 	wg.Add(2)
 
-	transfer := func(side string, dst, src io.ReadWriteCloser) {
-		_, err := io.Copy(dst, src)
+	transfer := func(side string, dst, src net.Conn) {
+		c.log.Debug("proxing %s -> %s", src.RemoteAddr(), dst.RemoteAddr())
+
+		n, err := io.Copy(dst, src)
 		if err != nil {
 			c.log.Debug("copy error: %s\n", err.Error())
```
Comment (cihangir) [inline client.go]: we could go with debug logs here as well like you did at line 412

File: virtualaddr.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
+package tunnel
+
+import (
```
Comment (cihangir) [inline virtualaddr.go]: can you clean up this file to have structs at the top and their respective functions grouped?

File: virtualaddr.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+		}
+
+		l.connCh <- conn
+	}
+}
+
+func (l *listener) localAddr() string {
+	if addr, ok := l.Addr().(*net.TCPAddr); ok {
+		if addr.IP.Equal(net.IPv4zero) {
+			return fmt.Sprintf("127.0.0.1:%d", addr.Port)
```
Comment (cihangir) [inline virtualaddr.go]: net.JoinHostPort

File: virtualaddr.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+		}
+	}
+	return l.Addr().String()
+}
+
+func (l *listener) stop() {
+	if atomic.CompareAndSwapInt32(&l.done, 0, 1) {
+		// self-pipe to break out of serve loop
+		conn, err := net.DialTimeout("tcp", l.localAddr(), defaultTimeout)
```
Comment (cihangir) [inline virtualaddr.go]: how does this stop the listener after all connections are closed/terminated?

also what does `self-pipe to break out of serve loop` mean?

File: client.go (hunk 1 above)
Comment (rjeczalik) [inline client.go]: done

File: virtualaddr.go (hunk 2 above)
Comment (rjeczalik) [inline virtualaddr.go]: done

File: virtualaddr.go (hunk 3 above)
Comment (rjeczalik) [inline virtualaddr.go]: done

File: virtualaddr.go (hunk 4 above)
Comment (rjeczalik) [inline virtualaddr.go]: `l.Accept()` will block until new connection arrives, we issue a dummy connect, so `l.Accept` returns and we return from the loop [here](https://github.com/koding/tunnel/pull/12/files#diff-9ff35f94bcd60257e5e3009930af187eR56). We do this as we can't just simply close the listener, as it is managed by the user (user may add / remove and the again add the listener multiple times, we're going to use this property in tunnelserver).

It's also safe if `l` is already busy accepting other connections, and we return from the loop after a user connection and not this one - either `DialTimeout` will timeout, or the dummy connection gets closed before we request yamux session.

ref: https://cr.yp.to/docs/selfpipe.html

Comment (cihangir) [inline virtualaddr.go]: can you write these as documentation?

Comment (rjeczalik) [inline virtualaddr.go]: done

**Comment by rjeczalik** [comment]
The failure [here](https://travis-ci.org/koding/tunnel/jobs/113538172) is an old flake - we're not waiting in tests for the yamux session to get established before issuing http requests (partially workarounded [in some tests](https://github.com/koding/tunnel/blob/87a88a1df17ab6a13130d87e043ff7348ef8f2ad/tunnel_test.go#L258-L259)). Will get that fixed in separate PR.

**Comment by cihangir** [comment]
there are two points that are not clear for me yet but, will check second pr with tests before commenting on them :)

**Comment by cihangir** [comment]
:+1:

**Pull request merged**